### PR TITLE
Minecraft Version support

### PR DESCRIPTION
**Pull request by KonoTyran**
~~small change to read java and forge versions from host.yaml to make it easier to test different versions of minecraft.~~

edit: now reads version info from https://github.com/KonoTyran/Minecraft_AP_Randomizer/blob/master/versions/minecraft_versions.json

### CODE REVIEW

**Comment by black-sliver** [comment]
Linux support (pointing host.yaml: minecraft_options: java to corretto17) still looks fine, but it get 

> No compatible mod version found for 1.18.2-40.1.0.

though with and without the --prerelease switch. Is this expected?

Another thing i realized while testing is that MinecraftClient does not validate the file argument is an apmc file, which could be added in a future version to get a useful error (currently you will get 'no .apmc file found." during server start).

**Comment by KonoTyran** [comment]
I'm silly and mixed up some variable names. as well as forgot to separate out the mc version from the forge string. all those issues are fixed now.

**Comment by black-sliver** [comment]
lgtm

File: MinecraftClient.py [hunk 1]
```diff
@@ -211,20 +197,20 @@ def install_forge(directory: str):
...
                 f.write(resp.content)
             print(f"Installing Forge...")
-            argstring = ' '.join([jdk, "-jar", "\"" + forge_install_jar+ "\"", "--installServer", "\"" + directory + "\""])
+            argstring = ' '.join([jdk, "-jar", "\"" + forge_install_jar + "\"", "--installServer", "\"" + directory + "\""])
             install_process = Popen(argstring, shell=not is_windows)
             install_process.wait()
             os.remove(forge_install_jar)
 
 
 # Run the Forge server. Return process object
```
Comment (Berserker66) [inline MinecraftClient.py:206]: Should technically be a docstring, the return could a be -> returntype

File: MinecraftClient.py [hunk 2]
```diff
@@ -242,46 +236,105 @@ def run_forge_server(forge_dir: str, heap_arg):
...
+    parser.add_argument('--release_channel', '-r', dest="channel", type=str, action='store',
+                        help="Specify release channel to use.")
+    parser.add_argument('--java', '-j', metavar='17', dest='java', type=str, default=False, action='store',
+                        help="specify java version.")
+    parser.add_argument('--forge', '-f', metavar='1.18.2-40.1.0', dest='forge', type=str, default=False, action='store',
+                        help="specify forge version. (Minecraft Version-Forge Version)")
+    parser.add_argument('--java', '-j', metavar='17', dest='java', type=str, default=False, action='store',
+                        help="specify java version.")
+    parser.add_argument('--forge', '-f', metavar='1.18.2-40.1.0', dest='forge', type=str, default=False, action='store',
+                        help="specify forge version. (Minecraft Version-Forge Version)")
```
Comment (black-sliver) [inline MinecraftClient.py:284]: `--java` and `--forge` is defined twice here - probably happened when merging conflicts from main?
```suggestion
    parser.add_argument('--java', '-j', metavar='17', dest='java', type=str, default=False, action='store',
                        help="specify java version.")
    parser.add_argument('--forge', '-f', metavar='1.18.2-40.1.0', dest='forge', type=str, default=False, action='store',
                        help="specify forge version. (Minecraft Version-Forge Version)")
```

File: MinecraftClient.py [hunk 3]
```diff
@@ -242,46 +236,105 @@ def run_forge_server(forge_dir: str, heap_arg):
...
-    if apmc_file is not None and not os.path.isfile(apmc_file):
-        raise FileNotFoundError(f"Path {apmc_file} does not exist or could not be accessed.")
+    if apmc_data is None:
+        raise FileNotFoundError(f"APMC file does not exist or is inaccessible at the given location ({apmc_file})")
+
+    if not os.path.isdir(java_dir) and is_windows:
+        if prompt_yes_no("Did not find java directory. Download and install java now?"):
+            download_java(java_version)
+        if not os.path.isdir(java_dir):
+            raise NotADirectoryError(f"Path {java_dir} does not exist or could not be accessed.")
```
Comment (black-sliver) [inline MinecraftClient.py:325]: This has multiple problems:
* The two parts of the first `if` should be swapped, so we don't check path on non-windows
* If java is missing, java_dir will be None leading to an exception in os.path.isdir
* After installation, java_dir is not updated, so the final `if` will also fail

File: MinecraftClient.py [hunk 4]
```diff
@@ -242,46 +236,105 @@ def run_forge_server(forge_dir: str, heap_arg):
...
+
+    if apmc_file is not None:
+        apmc_data = read_apmc_file(apmc_file)
+        data_version = apmc_data.get('client_version', '')
+
+    versions = get_minecraft_versions(data_version, channel)
+
     forge_dir = options["minecraft_options"]["forge_directory"]
     max_heap = options["minecraft_options"]["max_heap_size"]
+    forge_version = args.forge or versions["forge"]
```
Comment (black-sliver) [inline MinecraftClient.py:305]: Not sure if necessary, but there is now a way to detect wrong java and wrong mod, but no way to detect (and update) wrong forge.

Comment (KonoTyran) [reply MinecraftClient.py:305]: its absolutely necessary, and now fixed ^_^ thanks for catching that.

**Comment by black-sliver** [comment]
Ran into yet another problem :S
Hope you can reproduce it: When upgrading forge, the resulting install seems to be broken (terminates in an exception regarding chunk files). It does work however if I delete the `Minecraft Forge server` first to get a fresh install.

Other than that it seems to work nicely now.

**Comment by KonoTyran** [comment]
@black-sliver I an unable to reproduce this steps I used:

all these tests were done with a data version 7 APMC from the current website.
setup: no forge install, no java install in ap diretory.

1. use "release" channel in `host.yaml`
2. run minecraft client with version 7 APMC. let the it install `jdk16`, and forge `1.17.1-37.1.1`
3. connect to server walk around see it works.
4. switch to "beta" channel in `host.yaml`
5. run minecraft client with version 7 APMC let it install `jdk17`, and forge `1.18.2-40.1.0`

the server complains about missing keys (this is normal during an update as some things changed and shouldn't affect anything too much) but still continues to boot up and allow me to connect to it.

**Comment by black-sliver** [comment]
Switching back from beta to release triggered the problem for me

**Comment by KonoTyran** [comment]
ahh that's an issue with trying to downgrade a map from 1.18 to 1.17, as 1.18 has blocks above and below the minimum and maximum height for 1.17, downgrading worlds it not supported by MC itself. so not really a bug persay. and no way for me to fix. it.

**Comment by black-sliver** [comment]
So switching between beta and release may break the save (because it can't downgrade, you may end up breaking and regenerating the world), but if that switch is not supposed to work mid-seed, then that's totally fine. The users should just stay on release unless they know what they are doing anyway.

**Comment by KonoTyran** [comment]
I've also added a comment to `host.yaml` to clarify this behavior.

**Review by Berserker66 — APPROVED** [review]
Other than the form-error of using

```
# comment here
def function
```

instead of 
```
def function
   """comment here"""
```
   
Looks good. If you do change it to a docstring then IDEs will actually be able to display the text.

**Pull request merged**